borg/archiver.py

```diff
@@ -1399,7 +1440,7 @@ def build_parser(self, args=None, prog=None):
         subparser.add_argument('location', metavar='ARCHIVE',
                                type=location_validator(archive=True),
                                help='name of archive to create (must be also a valid directory name)')
-        subparser.add_argument('paths', metavar='PATH', nargs='+', type=str,
+        subparser.add_argument('paths', metavar='PATH', nargs='*', type=str,
```

Comment:
This is another approach to fix the argparse bug.
It uses the proposed fix from Steven Bethard and monkey patches the inner function `consume_positionals()` in a rather defensive way. (Revert all changes if a test fails after patching.)

```python
# https://bugs.python.org/issue15112
import argparse

# Thanks to Martijn Pieters
# http://stackoverflow.com/a/27550237/3456480
def monkeypatch_inner_function(outer, new_inner):
    """Replace a nested function code object used by outer with new_inner

    The replacement new_inner must use the same name and must at most use the
    same closures as the original.

    """
    if hasattr(new_inner, '__code__'):
        # support both functions and code objects
        new_inner = new_inner.__code__
    # find original code object so we can validate the closures match
    ocode = outer.__code__
    function, code = type(outer), type(ocode)
    iname = new_inner.co_name
    orig_inner = next(
        const for const in ocode.co_consts
        if isinstance(const, code) and const.co_name == iname)
    # you can ignore later closures, but since they are matched by position
    # the new sequence must match the start of the old.
    assert (orig_inner.co_freevars[:len(new_inner.co_freevars)] ==
            new_inner.co_freevars), 'New closures must match originals: %s'.format(orig_inner.co_freevars)
    # replace the code object for the inner function
    new_consts = tuple(
        new_inner if const is orig_inner else const
        for const in ocode.co_consts)
    # create a new function object with the new constants
    # note: python 2 code objects don't have co_kwonlyargcount
    c = code(ocode.co_argcount, ocode.co_kwonlyargcount, ocode.co_nlocals,
             ocode.co_stacksize, ocode.co_flags, ocode.co_code, new_consts,
             ocode.co_names, ocode.co_varnames, ocode.co_filename,
             ocode.co_name, ocode.co_firstlineno, ocode.co_lnotab,
             ocode.co_freevars, ocode.co_cellvars)
    return function(
        c,
        outer.__globals__, outer.__name__, outer.__defaults__,
        outer.__closure__)

def create_argparse_consume_positionals():
    # this is a copy of consume_positionals from python 3.6
    # with a proposed fix for bug https://bugs.python.org/issue15112
    # https://hg.python.org/cpython/file/3.6/Lib/argparse.py#l1916    
    # https://bugs.python.org/file26485/15112.patch
    self = None
    arg_strings_pattern = None
    positionals = None
    arg_strings = None
    take_action = None
    def consume_positionals(start_index):
        # match as many Positionals as possible
        match_partial = self._match_arguments_partial
        selected_pattern = arg_strings_pattern[start_index:]
        arg_counts = match_partial(positionals, selected_pattern)

        # if we haven't hit the end of the command line strings,
        # then don't consume any final zero-width arguments yet
        # (we may need to parse some more optionals first)
        if start_index + sum(arg_counts) != len(arg_strings_pattern):
            while arg_counts and arg_counts[-1] == 0:
                arg_counts.pop()

        # slice off the appropriate arg strings for each Positional
        # and add the Positional and its args to the list
        for action, arg_count in zip(positionals, arg_counts):
            args = arg_strings[start_index: start_index + arg_count]
            start_index += arg_count
            take_action(action, args)

        # slice off the Positionals that we just parsed and return the
        # index at which the Positionals' string args stopped
        positionals[:] = positionals[len(arg_counts):]
        return start_index
    return consume_positionals

def argparse_has_positional_bug():
    parser = argparse.ArgumentParser()
    parser.add_argument('--foo', action='store_true')
    parser.add_argument('pos1')
    arg = parser.add_argument('pos2', nargs='*')
    args, unrecognized = parser.parse_known_args("1 --foo 2".split())
    return True if unrecognized == [ '2' ] else False

def maybe_fix_argparse():
    if argparse_has_positional_bug():
        # print("has bug")
        consume_positionals = create_argparse_consume_positionals()
        origmethod = argparse.ArgumentParser._parse_known_args
        try:
            argparse.ArgumentParser._parse_known_args = monkeypatch_inner_function(
                origmethod, consume_positionals)
            if argparse_has_positional_bug():
                # print("still has bug")
                raise Exception()
            else:
                # print("fixed")
                return True
        except Exception:
            # fix didn't work..
            # strategic withdrawal!
            argparse.ArgumentParser._parse_known_args = origmethod
            return False
    else:
        # no need to fix
        return None



if __name__ == '__main__':
    import sys
    result = maybe_fix_argparse()
    print("{!r}".format(result))
```